primitives/arithmetic/src/lib.rs

```diff
@@ -85,8 +86,311 @@ where
 	}
 }
 
+/// A collection-like object that is made of values of type `T` and can normalize its individual
+/// values around a centric point.
+pub trait Normalizable<T> {
+	/// Normalize self around `t_max`.
+	///
+	/// Only returns `Ok` if the new sum of results is guaranteed to be equal to `t_max`. Else,
+	/// returns an error explaining why it failed to do so.
+	fn normalize(&self, t_max: T) -> Result<Vec<T>, &'static str>;
+}
+
+impl<T: Clone + Copy + Ord + BaseArithmetic + Debug> Normalizable<T> for Vec<T> {
+	fn normalize(&self, t_max: T) -> Result<Vec<T>, &'static str> {
+		normalize(self.as_ref(), t_max)
+	}
+}
+
+/// Normalize `input` so that the sum of all elements reaches `t_max`.
+///
+/// This implementation is currently in a balanced position between being performant and accurate.
+///
+/// 1. We prefer storing original indices, and sorting the `input` only once. This will save the
+///    cost of sorting per round at the cost of a little bit of memory.
+/// 2. The granularity of increment/decrements is determined by the difference and the number of
+///    elements in `input` and their sum difference with `t_max`. In short, the implementation is
+///    very likely to over-increment/decrement in case the `diff / input.len()` is rather small.
+///
+/// This function can return an error is if `T` cannot be built from the size of `input`, or if
+/// `sum(input)` cannot fit inside `T`. Moreover, if any of the internal operations saturate, it
+/// will also return en `Err`.
+///
+/// Based on the above, the best use case of the function will be only to correct small rounding
+/// errors. If the difference of the elements is vert large, then the subtraction/addition of one
+/// element with `diff / input.len()` might easily saturate and results will be `Err`.
+pub fn normalize<T>(input: &[T], t_max: T) -> Result<Vec<T>, &'static str>
```

Comment:
`t_max` should get a more concise name, semantically it's more of a `expected_input_sum`, iiuc?